--- a/sha256_alt.h
+++ b/sha256_alt.h
@@ -39,11 +39,16 @@ extern "C" {
 #define mbedtls_sha256_context ltc_hash_ctx_t
 
 #elif defined(MBEDTLS_FREESCALE_LPC_SHA256)
+#if (SECLIB_MODE_ALT)
+#include "SecLib_alt.h"
 
 /**
  * \brief          SHA-256 context structure
  */
+#define mbedtls_sha256_context sha256Context_t
+#else
 #define mbedtls_sha256_context sha_ctx_t
+#endif
 
 #elif defined(MBEDTLS_FREESCALE_CAAM_SHA256)
 
