diff --git a/components/wifi_bt_module/template/app_config.h b/components/wifi_bt_module/template/app_config.h
index c65f2a11..320a1a98 100644
--- a/components/wifi_bt_module/template/app_config.h
+++ b/components/wifi_bt_module/template/app_config.h
@@ -157,7 +157,7 @@
 #define WIFI_BT_USE_USD_INTERFACE
 #define WLAN_ED_MAC_CTRL                        \
     {                                           \
-        .ed_ctrl_2g = 0x1, .ed_offset_2g = 0x1E \
+        .ed_ctrl_2g = 0x1, .ed_offset_2g = 0x0E \
     }
 
 /* Murata 2DS */
@@ -173,7 +173,7 @@
 #define WIFI_BT_USE_M2_INTERFACE
 #define WLAN_ED_MAC_CTRL                        \
     {                                           \
-        .ed_ctrl_2g = 0x1, .ed_offset_2g = 0x1E \
+        .ed_ctrl_2g = 0x1, .ed_offset_2g = 0x0E \
     }
 
 /* Murata 1XK + Murata uSD-M.2 adapter */
